File: backend/esport-api/store.js
"use strict";

const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const { buildClientMatchList } = require("./match_merge");
const { formatOdds, formatBetOdds } = require("../shared/odds_format.js");

const DATA_DIR = process.env.ESPORT_DATA_DIR
  || path.join(__dirname, "..", "data", "esport");

function filePath(name) {
  return path.join(DATA_DIR, `${name}.json`);
}

function readJson(name, fallback) {
  const fp = filePath(name);
  try {
    if (!fs.existsSync(fp)) return fallback;
    return JSON.parse(fs.readFileSync(fp, "utf8"));
  } catch {
    return fallback;
  }
}

function writeJson(name, data) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
  fs.writeFileSync(filePath(name), JSON.stringify(data, null, 2), "utf8");
}

function hashPassword(password, salt) {
  return crypto.pbkdf2Sync(password, salt, 120000, 32, "sha256").toString("hex");
}

function newToken() {
  return crypto.randomBytes(24).toString("hex");
}

const DEFAULT_USER_KV = {
  CollectConfig: JSON.stringify({
    log: false,
    collect: [["OB", true], ["RAY", true]],
  }),
  USERCONFIG: JSON.stringify({
    betting: false,
    betMoney: 100,
    profit: 1.03,
    maxProfit: 1.2,
    minOdds: 1.3,
    maxOdds: 10,
    betSorting: "Custom",
  }),
  ACCOUNT: JSON.stringify([]),
  PROXY: JSON.stringify([]),
  GoogleCode: JSON.stringify([]),
  Wallet: JSON.stringify([]),
  Message: JSON.stringify({ telegramId: "", pushOrderId: "" }),
  Follow: JSON.stringify({}),
};

function ensureSeed() {
  const users = readJson("users", null);
  if (!users) {
    const salt = crypto.randomBytes(16).toString("hex");
    writeJson("users", [
      {
        id: 1,
        userName: "admin",
        passwordHash: hashPassword("admin", salt),
        salt,
        setting: {},
      },
    ]);
    writeJson("sessions", {});
    writeJson("platforms", {});
    writeJson("matches", {});
    writeJson("bets", {});
    writeJson("user_kv", { ...DEFAULT_USER_KV });
    writeJson("live_timers", {});
    writeJson("tag_platforms", {});
    writeJson("players", {});
    writeJson("money_logs", {});
    writeJson("player_orders", {});
    return;
  }
  const kv = readJson("user_kv", {});
  let changed = false;
  for (const [key, val] of Object.entries(DEFAULT_USER_KV)) {
    if (kv[key] == null) {
      kv[key] = val;
      changed = true;
    }
  }
  if (changed) writeJson("user_kv", kv);
}

function stableMatchId(gid) {
  let h = 0;
  for (const c of String(gid)) h = (Math.imul(31, h) + c.charCodeAt(0)) >>> 0;
  return h || 1;
}

function sourceFromBet(provider, b) {
  return {
    Type: provider,
    BetID: String(b.SourceBetID),
    HomeID: String(b.SourceHomeID),
    AwayID: String(b.SourceAwayID),
    HomeOdds: formatOdds(b.HomeOdds),
    AwayOdds: formatOdds(b.AwayOdds),
    Status: b.Status || "Normal",
  };
}

function getUserByName(userName) {
  const users = readJson("users", []);
  return users.find((u) => u.userName === userName);
}

function getUserById(userId) {
  const users = readJson("users", []);
  return users.find((u) => u.id === userId) || null;
}

function createUser(userName, password, setting = {}) {
  const users = readJson("users", []);
  const salt = crypto.randomBytes(16).toString("hex");
  const id = users.length ? Math.max(...users.map((u) => u.id)) + 1 : 1;
  const user = {
    id,
    userName,
    passwordHash: hashPassword(password, salt),
    salt,
    setting: {
      a8UserName: userName,
      a8Password: password,
      ...setting,
    },
  };
  users.push(user);
  writeJson("users", users);
  return user;
}

function updateUserSetting(userId, patch) {
  const users = readJson("users", []);
  const idx = users.findIndex((u) => u.id === userId);
  if (idx < 0) return null;
  users[idx].setting = { ...(users[idx].setting || {}), ...patch };
  writeJson("users", users);
  return users[idx];
}

function getSession(token) {
  if (!token) return null;
  const sessions = readJson("sessions", {});
  return sessions[token] || null;
}

function getUserByToken(token) {
  if (!token) return null;
  const sessions = readJson("sessions", {});
  const session = sessions[token];
  if (!session) return null;
  const users = readJson("users", []);
  return users.find((u) => u.id === session.userId) || null;
}

function createSession(userId, extra = {}) {
  const token = newToken();
  const sessions = readJson("sessions", {});
  sessions[token] = { userId, createdAt: Date.now(), ...extra };
  writeJson("sessions", sessions);
  return token;
}

function removeSession(token) {
  const sessions = readJson("sessions", {});
  delete sessions[token];
  writeJson("sessions", sessions);
}

function getPlatform(provider) {
  const platforms = readJson("platforms", {});
  return platforms[provider] || null;
}

function setPlatform(provider, data) {
  const platforms = readJson("platforms", {});
  platforms[provider] = { ...platforms[provider], ...data, provider, updatedAt: Date.now() };
  writeJson("platforms", platforms);
  return platforms[provider];
}

function pruneBetsForProvider(provider, activeMatchIds) {
  const active = new Set(activeMatchIds.map(String));
  const all = readJson("bets", {});
  const prefix = `${provider}:`;
  let changed = false;
  for (const key of Object.keys(all)) {
    if (!key.startsWith(prefix)) continue;
    const matchId = key.slice(prefix.length);
    if (!active.has(matchId)) {
      delete all[key];
      changed = true;
    }
  }
  if (changed) writeJson("bets", all);
}

/** 每次 SaveMatch 用本批列表覆盖该 provider 下的全部比赛（平台当前快照） */
function saveMatches(provider, matchs) {
  const all = readJson("matches", {});
  const now = Date.now();
  const next = {};
  const list = Array.isArray(matchs) ? matchs : [];
  for (const m of list) {
    if (!m || m.SourceMatchID == null) continue;
    next[String(m.SourceMatchID)] = { ...m, provider, savedAt: now };
  }
  all[provider] = next;
  writeJson("matches", all);
  pruneBetsForProvider(provider, Object.keys(next));
}

function mergeBetsByMap(existing, incoming) {
  const byMap = new Map();
  for (const b of existing || []) {
    byMap.set(Number(b.Map ?? 0), formatBetOdds(b));
  }
  for (const b of incoming || []) {
    byMap.set(Number(b.Map ?? 0), formatBetOdds(b));
  }
  return [...byMap.values()].sort((a, b) => (a.Map ?? 0) - (b.Map ?? 0));
}

/**
 * A8 采集按场/按 stage 多次上传；OB 分 stage 轮询时单批常不完整。
 * 按 Map 合并，空批次不覆盖，避免地图行时有时无。
 */
function saveBets(provider, matchId, bets) {
  const all = readJson("bets", {});
  const key = `${provider}:${matchId}`;
  const existing = all[key]?.bets || [];
  const incoming = Array.isArray(bets) ? bets : [];

  if (incoming.length === 0) {
    if (existing.length > 0) return;
    all[key] = { provider, matchId: String(matchId), bets: [], savedAt: Date.now() };
    writeJson("bets", all);
    return;
  }

  const merged = mergeBetsByMap(existing, incoming);
  all[key] = { provider, matchId: String(matchId), bets: merged, savedAt: Date.now() };
  writeJson("bets", all);
}

function saveLiveTimer(provider, timer) {
  const all = readJson("live_timers", {});
  all[provider] = { provider, timer, savedAt: Date.now() };
  writeJson("live_timers", all);
}

function getUserKv(key) {
  const kv = readJson("user_kv", {});
  return kv[key] ?? null;
}

function setUserKv(key, content) {
  const kv = readJson("user_kv", {});
  kv[key] = content;
  writeJson("user_kv", kv);
}

function parseKvContent(raw) {
  if (raw == null || raw === "") return null;
  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

function buildMatchList() {
  return buildClientMatchList({
    matches: readJson("matches", {}),
    bets: readJson("bets", {}),
    timers: readJson("live_timers", {}),
    sourceFromBet,
  });
}

module.exports = {
  DATA_DIR,
  ensureSeed,
  getUserByName,
  getUserById,
  getUserByToken,
  createUser,
  updateUserSetting,
  getSession,
  createSession,
  removeSession,
  hashPassword,
  getPlatform,
  setPlatform,
  saveMatches,
  saveBets,
  saveLiveTimer,
  getUserKv,
  setUserKv,
  parseKvContent,
  buildMatchList,
  readJson,
  writeJson,
};

File: gamebet_backend_python/esport_api/store.py
import json
import os
import threading
import hashlib
import secrets
import time
from pathlib import Path

from shared.storage_paths import ESPORT_DATA_DIR
from shared.match_time import a8_start_time_list_allowed, normalize_epoch_ms
from shared.im_parse import normalize_im_bet
from shared.odds_format import format_bet_odds
import db.store as db_store

DATA_DIR = ESPORT_DATA_DIR
CLIENT_MATCHS_DEBOUNCE_MS = int(os.environ.get("ESPORT_CLIENT_MATCHS_DEBOUNCE_MS", "400")) / 1000

_lock = threading.Lock()
_rebuild_timer: threading.Timer | None = None

DEFAULT_USER_KV = {
    "CollectConfig": json.dumps({"log": False, "collect": []}),
    "USERCONFIG": json.dumps({"betting": False, "betMoney": 100, "profit": 1.03,
                               "maxProfit": 1.2, "minOdds": 1.3, "maxOdds": 10, "betSorting": "Custom"}),
    "ACCOUNT": json.dumps([]),
    "PROXY": json.dumps([]),
    "GoogleCode": json.dumps([]),
    "Wallet": json.dumps([]),
    "Message": json.dumps({"telegramId": "", "pushOrderId": ""}),
    "Follow": json.dumps({}),
}
USER_SETTING_KEYS = [k for k in DEFAULT_USER_KV if k != "ACCOUNT"]


def _file_path(name: str) -> Path:
    return DATA_DIR / f"{name}.json"


def read_json(name: str, fallback=None):
    fp = _file_path(name)
    try:
        if not fp.exists():
            return fallback
        return json.loads(fp.read_text(encoding="utf-8"))
    except Exception:
        return fallback


def write_json(name: str, data):
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    _file_path(name).write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def hash_password(password: str, salt: str) -> str:
    return hashlib.pbkdf2_hmac("sha256", password.encode(), salt.encode(), 120000).hex()


def new_token() -> str:
    return secrets.token_hex(24)


def _now_ms() -> int:
    return int(time.time() * 1000)


# ── seed ──────────────────────────────────────────────────────────────────────

def ensure_seed():
    users = read_json("users", None)
    if isinstance(users, list) and users:
        db_store.save_users(users)
    if not db_store.list_users():
        salt = secrets.token_hex(16)
        default_user = {
            "id": 1, "userName": "admin",
            "passwordHash": hash_password("admin", salt),
            "salt": salt, "setting": {},
        }
        db_store.save_users([default_user])
        if not users:
            write_json("users", [default_user])
        write_json("sessions", {})
        write_json("platforms", {})
        write_json("matches", {})
        write_json("bets", {})
        write_json("user_kv", dict(DEFAULT_USER_KV))
        write_json("live_timers", {})
        write_json("client_matchs", {"builtAt": 0, "count": 0, "info": []})
        write_json("tag_platforms", {})
        write_json("players", {})
        write_json("money_logs", {})
        write_json("player_orders", {})
        return

    kv = read_json("user_kv", {})
    changed = False
    for key, val in DEFAULT_USER_KV.items():
        if kv.get(key) is None:
            kv[key] = val
            changed = True
    if changed:
        write_json("user_kv", kv)

    if db_store.count_accounts() == 0:
        tj_user = get_user_by_name("TJ01")
        legacy_accounts = _parse_kv_content(kv.get("ACCOUNT"))
        if tj_user and isinstance(legacy_accounts, list) and legacy_accounts:
            db_store.replace_accounts_for_user(tj_user["id"], legacy_accounts)

    if db_store.count_user_settings() == 0:
        tj_user = get_user_by_name("TJ01")
        if tj_user:
            for key in USER_SETTING_KEYS:
                db_store.set_user_setting(tj_user["id"], key, kv.get(key) or DEFAULT_USER_KV.get(key, ""))


# ── users ─────────────────────────────────────────────────────────────────────

def get_user_by_name(user_name: str) -> dict | None:
    return db_store.get_user_by_name(user_name)


def get_user_by_id(user_id: int) -> dict | None:
    return db_store.get_user_by_id(user_id) or None


def create_user(user_name: str, password: str, setting: dict = None) -> dict | None:
    salt = secrets.token_hex(16)
    user = {
        "userName": user_name,
        "passwordHash": hash_password(password, salt),
        "salt": salt,
        "setting": {"a8UserName": user_name, "a8Password": password, **(setting or {})},
    }
    return db_store.create_user(user)


def update_user_setting(user_id: int, patch: dict) -> dict | None:
    return db_store.update_user_setting(user_id, patch)


def get_user_by_token(token: str) -> dict | None:
    if not token:
        return None
    sessions = read_json("sessions", {})
    session = sessions.get(token)
    if not session:
        return None
    return get_user_by_id(session.get("userId"))


def get_session(token: str):
    if not token:
        return None
    return read_json("sessions", {}).get(token)


def create_session(user_id: int, extra: dict = None) -> str:
    token = new_token()
    sessions = read_json("sessions", {})
    sessions[token] = {"userId": user_id, "createdAt": _now_ms(), **(extra or {})}
    write_json("sessions", sessions)
    return token


def remove_session(token: str):
    sessions = read_json("sessions", {})
    sessions.pop(token, None)
    write_json("sessions", sessions)


# ── platform ──────────────────────────────────────────────────────────────────

def get_platform(provider: str):
    return read_json("platforms", {}).get(provider)


def set_platform(provider: str, data: dict) -> dict:
    platforms = read_json("platforms", {})
    platforms[provider] = {**(platforms.get(provider) or {}), **data,
                           "provider": provider, "updatedAt": _now_ms()}
    write_json("platforms", platforms)
    return platforms[provider]


# ── matches / bets ────────────────────────────────────────────────────────────

def _prune_bets_for_provider(provider: str, active_match_ids: list):
    active = set(str(i) for i in active_match_ids)
    all_bets = read_json("bets", {})
    prefix = f"{provider}:"
    changed = False
    for key in list(all_bets.keys()):
        if not key.startswith(prefix):
            continue
        match_id = key[len(prefix):]
        if match_id not in active:
            del all_bets[key]
            changed = True
    if changed:
        write_json("bets", all_bets)


def save_matches(provider: str, matchs: list):
    all_matches = read_json("matches", {})
    now = _now_ms()
    next_matches: dict = {}
    lst = matchs if isinstance(matchs, list) else []
    prev = all_matches.get(provider)
    if not lst and prev and isinstance(prev, dict) and prev:
        return
    for m in lst:
        if not m or m.get("SourceMatchID") is None:
            continue
        start = normalize_epoch_ms(m.get("StartTime") or 0)
        if start > 0 and not a8_start_time_list_allowed(start):
            continue
        next_matches[str(m["SourceMatchID"])] = {**m, "provider": provider, "savedAt": now}
    all_matches[provider] = next_matches
    write_json("matches", all_matches)
    _prune_bets_for_provider(provider, list(next_matches.keys()))
    rebuild_client_match_list_now()


def _merge_bets_by_map(existing: list, incoming: list) -> list:
    by_map: dict = {}
    for b in (existing or []):
        name = str(b.get("BetName") or "")
        if "+" in name:
            continue
        by_map[int(b.get("Map") or 0)] = format_bet_odds(b)
    for b in (incoming or []):
        name = str(b.get("BetName") or "")
        if "+" in name:
            continue
        by_map[int(b.get("Map") or 0)] = format_bet_odds(b)
    return sorted(by_map.values(), key=lambda b: b.get("Map") or 0)


def save_bets(provider: str, match_id, bets):
    all_bets = read_json("bets", {})
    key = f"{provider}:{match_id}"
    existing = (all_bets.get(key) or {}).get("bets") or []
    raw = bets if isinstance(bets, list) else []
    incoming = [normalize_im_bet(b) for b in raw] if provider == "IM" else raw

    if not incoming:
        if existing:
            return
        all_bets[key] = {"provider": provider, "matchId": str(match_id), "bets": [], "savedAt": _now_ms()}
        write_json("bets", all_bets)
        schedule_rebuild_client_match_list()
        return

    merged = _merge_bets_by_map(existing, incoming)
    all_bets[key] = {"provider": provider, "matchId": str(match_id), "bets": merged, "savedAt": _now_ms()}
    write_json("bets", all_bets)
    schedule_rebuild_client_match_list()


def save_live_timer(provider: str, timer):
    all_timers = read_json("live_timers", {})
    all_timers[provider] = {"provider": provider, "timer": timer, "savedAt": _now_ms()}
    write_json("live_timers", all_timers)
    schedule_rebuild_client_match_list()


# ── user kv / settings ────────────────────────────────────────────────────────

def get_user_kv(key: str):
    return read_json("user_kv", {}).get(key)


def set_user_kv(key: str, content: str):
    kv = read_json("user_kv", {})
    kv[key] = content
    write_json("user_kv", kv)


def is_user_setting_key(key: str) -> bool:
    return key in USER_SETTING_KEYS


def get_user_setting(user_id: int, key: str):
    if not is_user_setting_key(key):
        return None
    return db_store.get_user_setting(user_id, key)


def set_user_setting(user_id: int, key: str, content: str):
    if not is_user_setting_key(key):
        set_user_kv(key, content)
        return
    db_store.set_user_setting(user_id, key, content or "")


# ── accounts ──────────────────────────────────────────────────────────────────

def get_accounts_for_user(user_id: int) -> list:
    return db_store.list_accounts_for_user(user_id)


def set_accounts_for_user(user_id: int, accounts: list) -> list:
    return db_store.replace_accounts_for_user(user_id, accounts)


def update_account_for_user(user_id: int, account_id, updates: dict):
    return db_store.update_account_for_user(user_id, account_id, updates)


def remove_account_for_user(user_id: int, account_id) -> bool:
    return db_store.remove_account_for_user(user_id, account_id)


# ── match list ────────────────────────────────────────────────────────────────

def rebuild_client_match_list_now() -> list:
    from esport_api.match_merge import build_client_match_list
    from esport_api.default_odds import record_from_match_list
    matches = read_json("matches", {})
    bets = read_json("bets", {})
    timers = read_json("live_timers", {})
    info = build_client_match_list(matches, bets, timers)
    write_json("client_matchs", {"builtAt": _now_ms(), "count": len(info), "info": info})
    db_store.save_client_matches(info)
    db_store.prune_client_matches([int(m["ID"]) for m in info])
    record_from_match_list(info)
    return info


def schedule_rebuild_client_match_list():
    global _rebuild_timer
    with _lock:
        if _rebuild_timer:
            _rebuild_timer.cancel()
        _rebuild_timer = threading.Timer(CLIENT_MATCHS_DEBOUNCE_MS, _do_rebuild)
        _rebuild_timer.daemon = True
        _rebuild_timer.start()


def _do_rebuild():
    global _rebuild_timer
    with _lock:
        _rebuild_timer = None
    try:
        rebuild_client_match_list_now()
    except Exception as e:
        print(f"[store] rebuild client_matchs failed: {e}")


def build_match_list() -> list:
    cached = read_json("client_matchs", None)
    if cached and isinstance(cached.get("info"), list):
        return cached["info"]
    return rebuild_client_match_list_now()


def _parse_kv_content(raw):
    if raw is None or raw == "":
        return None
    try:
        return json.loads(raw)
    except Exception:
        return raw
